packages/band-arithmetic/band-arithmetic.js

```diff
@@ -0,0 +1,140 @@
+
+  const variable_index = variables.findIndex(variable => variable === input.name);
+  return band_values[variable_index];
+};
+
+const compute_value = (node, band_values) => {
+  const left_node = node.args[0];
+  const right_node = node.args[1];
+
+  const operation = node.fn;
+
+  let left_value;
+  if (left_node.content) { // if the node represents parentheses, it will be an object with a single property "content" which contains a node
+    left_value = compute_value(left_node.content, band_values);
+  } else if (is_leaf_node(left_node)) {
+    left_value = get_value(left_node, band_values);
+  } else {
+    left_value = compute_value(left_node, band_values);
+  }
+
+  let right_value;
+  if (right_node.content) { // if the node represents parentheses, it will be an object with a single property "content" which contains a node
+    right_value = compute_value(right_node.content, band_values);
+  } else if (is_leaf_node(right_node)) {
+    right_value = get_value(right_node, band_values);
+  } else {
+    right_value = compute_value(right_node, band_values);
+  }
+
+  return operations[operation](left_value, right_value);
+};
+
+const get_band_rows = (bands, index) => {
+  // using a for loop here instead of map leads to a significant performance improvement
+  const band_rows = [];
+  for (let i = 0; i < bands.length; i++) {
+    band_rows.push(bands[i][index]);
+  }
+  return band_rows;
+};
+
+const get_band_values = (band_rows, index) => {
+  // using a for loop here instead of map leads to a significant performance improvement
+  const band_values = [];
+  for (let i = 0; i < band_rows.length; i++) {
+    band_values.push(band_rows[i][index]);
+  }
+  return band_values;
+}
+
+// pre-parse arithmetic string to catch limitations with arithmetic operations
+// before attempting to compute
+const arithmetic_error = (arithmetic) => {
+  if (arithmetic.match(regex_multi_character)) {
+    return ('Geoblaze does not currently support implicit multiplication between variables. Please use the multiplication (*) symbol for these operations.');
+  }
+}
+
+/**
+ * The band arithmetic function takes a raster and an arithmetic operation written as
+ * a string as input. The function performs pixel-by-pixel calculation according to the
+ * arithmetic operation provided. This is only possible for a multiband raster and not
+ * for single band rasters. The output is a computed single band raster.
+ * @name band_arithmetic
+ * @param {Object} raster - a raster from the georaster library
+ * @param {String} operation - a string representation of a arithmetic operation to perform
+ * @returns {Object} array of computed values for each band
+ * @example
+ * const ndvi = geoblaze.band_arithmetic(georaster, '(c - b)/(c + b)');
+ */
+
+module.exports = (georaster, arithmetic) => {
+  return new Promise((resolve, reject) => {
+    const parsed_arithmetic = parse(arithmetic.toLowerCase());
+
+    if (georaster.values.length < 2) {
```

Comment:
@sdpeyton , It might not be a popular use case, but shouldn't we still allow people to operate on a raster?  For example, what if someone wanted to stretch their values on a single band raster?  Or wanted to convert their counts to density?

Reply:
Unfortunately, stretching doesn't work the same way as band arithmetic. I guess someone could multiply values by a constant but I don't believe there is much value gained in doing that. Typically band arithmetic is only used when you are trying to derive an output across multiple bands.